File: ios/Plugin/StripePlugin.swift
import Foundation
import Capacitor
import StripeApplePay
import PassKit

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(StripePlugin)
public class StripePlugin: CAPPlugin, ApplePayContextDelegate {
    private var payCallId: String?
    private var appleClientSecret: String?

    @objc func initialize(_ call: CAPPluginCall) {
        guard let publicKey = call.options["publicKey"] as? String, !publicKey.isEmpty else {
            call.reject("Call must provide \"publicKey\"")
            return
        }
        StripeAPI.defaultPublishableKey = publicKey
        STPAPIClient.shared.appInfo = STPAppInfo(name: "@dicty/stripe", partnerId: nil, version: nil, url: nil)
        call.resolve()
    }

    @objc func isApplePayAvailable(_ call: CAPPluginCall) {
        if (!StripeAPI.deviceSupportsApplePay()) {
            call.reject("Can not use on this Device.")
            return
        }

        call.resolve()
    }

    @objc func handleApplePay(_ call: CAPPluginCall) {
        guard let appleClientSecret = call.options["clientSecret"] as? String, !appleClientSecret.isEmpty else {
            call.reject("Call must provide \"clientSecret\"")
            return
        }
        guard let merchantIdentifier = call.options["merchantIdentifier"] as? String, !merchantIdentifier.isEmpty else {
            call.reject("Call must provide \"merchantIdentifier\"")
            return
        }
        guard let label = call.options["label"] as? String, !label.isEmpty else {
            call.reject("Call must provide \"label\"")
            return
        }
        guard let countryCode = call.options["countryCode"] as? String, !countryCode.isEmpty else {
            call.reject("Call must provide \"countryCode\"")
            return
        }
        guard let currency = call.options["currency"] as? String, !currency.isEmpty else {
            call.reject("Call must provide \"currency\"")
            return
        }
        guard let amount = call.options["amount"] as? Double, !amount.isLessThanOrEqualTo(0) else {
            call.reject("Call must provide \"amount\"")
            return
        }
        guard let billingPeriod = call.options["billingPeriod"] as? Double, !billingPeriod.isLessThanOrEqualTo(0) else {
            call.reject("Call must provide \"billingPeriod\"")
            return
        }
        guard let description = call.options["description"] as? String, !description.isEmpty else {
            call.reject("Call must provide \"description\"")
            return
        }
        guard let managementURL = call.options["managementURL"] as? String, !managementURL.isEmpty else {
            call.reject("Call must provide \"managementURL\"")
            return
        }
        guard let billingAgreement = call.options["billingAgreement"] as? String, !billingAgreement.isEmpty else {
            call.reject("Call must provide \"billingAgreement\"")
            return
        }

        self.appleClientSecret = appleClientSecret
        let request = StripeAPI.paymentRequest(withMerchantIdentifier: merchantIdentifier, country: countryCode, currency: currency)

        let billing = PKRecurringPaymentSummaryItem(label: label, amount: NSDecimalNumber(floatLiteral: amount))
        billing.startDate = Date()
        billing.endDate = Date().addingTimeInterval(billingPeriod)
        billing.intervalUnit = .month

        if #available(iOS 16.0, *) {
            request.recurringPaymentRequest = PKRecurringPaymentRequest(
                    paymentDescription: description,
                    regularBilling: billing,
                    managementURL: URL(string: managementURL)!
            )
            request.recurringPaymentRequest?.billingAgreement = billingAgreement
        } else {
            // Fallback on earlier versions
        }

        request.paymentSummaryItems = [billing]

        if let applePayContext = STPApplePayContext(paymentRequest: request, delegate: self) {
            DispatchQueue.main.async {
                self.bridge?.saveCall(call)
                self.payCallId = call.callbackId;
                applePayContext.presentApplePay()
            }
        } else {
            call.reject("STPApplePayContext is failed")
        }
    }

    @objc func isGooglePayAvailable(_ call: CAPPluginCall) {
        call.unavailable("Not implemented on iOS")
    }

    @objc func handleGooglePay(_ call: CAPPluginCall) {
        call.unavailable("Not implemented on iOS")
    }
}


extension StripePlugin {
    public func applePayContext(_ context: STPApplePayContext, didCreatePaymentMethod paymentMethod: StripeAPI.PaymentMethod, paymentInformation: PKPayment, completion: @escaping STPIntentClientSecretCompletionBlock) {
        let clientSecret = appleClientSecret;
        completion(clientSecret, nil);
    }

    public func applePayContext(_ context: STPApplePayContext, didCompleteWith status: STPApplePayContext.PaymentStatus, error: Error?) {
        if let callId = payCallId, let call = bridge?.savedCall(withID: callId) {
            payCallId = nil;
            appleClientSecret = nil;
            switch status {
            case .success:
                call.resolve(["result": ApplePayEvents.Completed.rawValue])
                break
            case .error:
                call.resolve(["result": ApplePayEvents.Failed.rawValue])
                break
            case .userCancellation:
                call.resolve(["result": ApplePayEvents.Cancelled.rawValue])
                break
            }
        }
    }
}

File: android/src/main/java/app/dicty/plugins/stripe/StripePlugin.java
package app.dicty.plugins.stripe;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.stripe.android.PaymentConfiguration;
import com.stripe.android.googlepaylauncher.GooglePayLauncher;

import org.jetbrains.annotations.NotNull;

@CapacitorPlugin(name = "Stripe")
public class StripePlugin extends Plugin {
    private GooglePayLauncher googlePayLauncher;
    private boolean isAvailable;
    private String payCallId;

    @Override
    public void load() {
        MetaData metaData = new MetaData(this::getContext);
        PaymentConfiguration.init(getContext(), metaData.publicKey, metaData.stripeAccount);

        googlePayLauncher = new GooglePayLauncher(
                getActivity(),
                new GooglePayLauncher.Config(metaData.env, metaData.countryCode, metaData.displayName),
                (boolean isReady) -> this.isAvailable = isReady,
                this::onGooglePayResult
        );
    }

    @PluginMethod
    public void initialize(final PluginCall call) {
        call.resolve();
    }

    private void onGooglePayResult(@NotNull GooglePayLauncher.Result result) {
        PluginCall call = bridge.getSavedCall(payCallId);
        payCallId = null;
        if (result instanceof GooglePayLauncher.Result.Completed) {
            // Payment succeeded, show a receipt view
            call.resolve(new JSObject().put("result", GooglePayEvents.Completed.getWebEventName()));
        } else if (result instanceof GooglePayLauncher.Result.Canceled) {
            call.resolve(new JSObject().put("result", GooglePayEvents.Cancelled.getWebEventName()));
        } else if (result instanceof GooglePayLauncher.Result.Failed) {
            call.resolve(new JSObject().put("result", GooglePayEvents.Failed.getWebEventName()));
        }
    }

    @PluginMethod
    public void isGooglePayAvailable(final PluginCall call) {
        if (isAvailable) {
            call.resolve();
            return;
        }
        call.unimplemented("Not implemented on Device.");
    }

    @PluginMethod
    public void handleGooglePay(final PluginCall call) {
        String clientSecret = call.getString("clientSecret");

        if (clientSecret == null) {
            call.reject("Call must provide \"clientSecret\"");
            return;
        }

        payCallId = call.getCallbackId();
        bridge.saveCall(call);
        googlePayLauncher.presentForPaymentIntent(clientSecret);
        call.resolve();
    }

    @PluginMethod
    public void isApplePayAvailable(final PluginCall call) {
        call.unavailable("Not implemented on Android");
    }

    @PluginMethod
    public void handleApplePay(final PluginCall call) {
        call.unavailable("Not implemented on Android");
    }
}
